src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -134,87 +134,301 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
                 return CreateChangeTypeArgumentException(srcEEType, dstEEType);
             }
 
+            EETypeElementType srcElementType = srcEEType->ElementType;
+
+            // This switch supports the conversions present in the table below, in 'CanPrimitiveWiden'
             switch (dstElementType)
             {
+                // Boolean (W = BOOL)
                 case EETypeElementType.Boolean:
-                    dstObject = Convert.ToBoolean(srcObject);
+
+                    // Can only be bool here
+                    dstObject = srcObject;
                     break;
 
+                // Char (W = U2, CHAR, I4, U4, I8, U8, R4, R8)
                 case EETypeElementType.Char:
-                    char charValue = Convert.ToChar(srcObject);
+                    char charValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.UInt16:
+                            charValue = (char)(ushort)srcObject;
```

Comment:
@jkotas was thinking, because we're already manually checking all types here, what about replacing all these checked unbox ops with just eg. here `Unsafe.As<byte, char>(ref RuntimeHelpers.GetRawData(srcObject))`? Or is there some other fast "unchecked unbox" internal helper I could use here?

Reply:
This method does low-level stuff already, so we can also make it go all the way by using GetRawData.
 
Also, I think you can make the code smaller by having just a single call of the boxing helper at very end. Something like:
```
ref byte dstValue;
...
switch
    switch
    switch
...
return RuntimeHelpers.Box(ref dstValue, dstEEType);
```

Reply:
> "I think you can make the code smaller by having just a single call of the boxing helper at very end"

Mmh not entirely sure I'm following how I'd declare the shared storage there, since I don't have a reference. Should I just declare a local with the size of the largest possible type, write there and then just reinterpret at the end? Ie.:

```csharp
long rawDestValue;
...
switch
    switch
        ...
        *(byte*)rawDestValue = theByte;
    switch
        ...
        *(short*)rawDestValue = theShort;
...
return RuntimeHelpers.Box(ref *(byte*)&rawDestValue, dstEEType);
```

Reply:
I was thinking more something like this:
```
void* dstValue;
byte theByte;
short theShort;
...
switch
    switch
        ...
        dstValue = &theByte;
    switch
        ...
        dstValue = &theShort;
...
return RuntimeHelpers.Box(ref *(byte*)dstValue, dstEEType);
```

Your sketch would work too. If you prefer to go with it, make sure to handle both alignment and endianness correctly. (E.g. `long` is not guaranteed to have sufficient alignment to store double on Arm32.)

Reply:
Thank you! I used your approach as it seemed simpler. Done in a2e50382b782cfb3a28a2d13c9bd0628a82288e8 🙂